File: tree/Traversal/Preorder/iterative/cpp/PreorderTraversal.cpp
#include "PreorderTraversal.hpp"

#include <vector>

using namespace std;

vector<int> PreorderTraversal(const vector<int>& leftChildren, const vector<int>& rightChildren, int root) {
    vector<int> order;
    vector<int> stack = {root};

    while (!stack.empty()) {
        const int node = stack.back();
        stack.pop_back();
        order.push_back(node);

        if (rightChildren[node] != -1) {
            stack.push_back(rightChildren[node]);
        }

        if (leftChildren[node] != -1) {
            stack.push_back(leftChildren[node]);
        }
    }

    return order;
}

File: tree/Traversal/Preorder/iterative/rust/src/PreorderTraversal.rs
#[allow(non_snake_case)]
pub fn PreorderTraversal(left_children: &[i32], right_children: &[i32], root: usize) -> Vec<usize> {
    let mut order = Vec::new();
    let mut stack = vec![root];

    while let Some(node) = stack.pop() {
        order.push(node);

        let right_child = right_children[node];
        if right_child != -1 {
            stack.push(right_child as usize);
        }

        let left_child = left_children[node];
        if left_child != -1 {
            stack.push(left_child as usize);
        }
    }

    order
}
